### PR TITLE
Feature: Parallax stars/haze while flying

### PR DESCRIPTION
**Pull request by Terin**
**Note:** This PR includes edits to multiple files for debug/testing for aesthetic preferences, which will be removed as the draft is finalized. The majority of final code is located in `source/StarField.cpp`, and code review should be focused on there.

## Feature Details
This PR adds the ability for the stars, haze, and main layers of the game to be drawn at different zoom levels, giving a parallax effect of depth to the player. The current stars and haze moving perfectly in line with the ships and planets provides a very flat look to the game, and even splitting the game into three basic layers goes a long way to addressing that. It also provides an option to disable it, in case it causes discomfort in the player.

Also included with this PR are keybinds and functions to allow the user to set the zoom values for the star and haze layers from the settings menu, and in real-time, rather than having to tweak a value in code and recompile. This should let reviewers and  Github users easily experiment before we settle on final values. I have it default to a 90% zoom for haze, and 70% zoom for stars, but perhaps them being further apart would look better, or even having the haze at a >100% value. Feel free to experiment and post what you think.

A future PR (or maybe a later commit in this one) would ideally provide further depth to this system: 3-4 layers for stars, a couple for haze, maybe even have stellar objects be slightly askew (although it could look for non-flagships landing). But even as-is, it seems to be a high-impact change to the aesthetics of Endless Sky.

## Example Video
https://cdn.discordapp.com/attachments/266345072554016768/808498951631142943/parallax3.mp4

## Testing Done
Flying around in space at multiple zoom levels, changing zoom levels in flight, and with layers at >100% zoom, to check that they move consistently.

## Performance Impact
Depending on what zoom value we use for the stars layer, there will be more stars added that could somewhat impact performance. On my Acer Chromebook C720, I notice a 1% increase to GPU usage when zoomed all the way out, on a 1366x768 display, with a 70% zoom value for the stars. There is also the option to disable parallax, if a user is trying to get as much performance out of their game as possible.

## PR Checklist
 - [x] Removed patch of #5714, temporarily provided here as it is crucial to testing the aesthetics of this feature
 - [x] Removed parallax-tweaking keybinds and functions, replaced with agreed-upon zoom values

### CODE REVIEW

**Comment by beccabunny** [comment]
Beautiful!

**Comment by beccabunny** [comment]
After testing, something I think is needed directly in this PR would be the different star layers, with 70% zoom being the lowest value going up to 85%ish. Still, really nice

**Comment by Zitchas** [comment]
I really enjoyed the videos you posted of this, I'm looking forward to trying it out and playing around with some values.

**Comment by Zitchas** [comment]
I just tried maxing out a bunch of variations:
TMP factors to 200 and.... It feels like the planets are chasing me. This is not the intended effect, I know. It's kind of trippy, actually. And looks really neat. 

Star 25, Haze 200 : I love the effect of having the haze stream by while the stars don't move much. Downside, the stars are very obviously tiled and repeating.

Star 200, Haze 25: Uhhh... Good way to watch haze tiling in. Stars move so fast that I feel like I should be arriving in the next system any time now.

Star 100, Haze 100: Looks like normal.

Star 50, Haze 75: This looks good, although I think I can pick out some tiling. Pretty good ratio, though.

Star 70, Haze 125: OK, I love this combo. I mean, really. I love this. The Ember Waste looks *awesome* like this, I just wish that the haze was *above* the ship for it. This really makes me wish we could have two layers of haze, one red haze at 75 and underneath (the way things are now), and a second haze at 125 on top. Now, the caveat, of course, is that I don't think the other hazes look as good with it. Basically, hazes that are supposed to be "stuff in the distance" look better with <100 zoom; but hazes that are supposed to be all around you look great with >100.

OK, wishlist time:
a) would it be possible to have the parallax for hazes set in the system attribute instead of the settings?
b) Could we have support for multiple layers of hazes with different amounts of parallax, again, set in the system attribute?

I think my preferred numbers would be something along the lines of:
Stars at something like 5, 20, 50. (Could be static, universal.)
Hazes set in the system file, supporting at least two layers with the option to be on top or underneath. The Ember Waste for instance, with two layers, one underneath at 75 and one overtop at 125.

**Comment by Ferociousfeind** [comment]
Is it beyond the scope of this PR to make all system elements move with different speeds (i.e. putting the planets and stars a little bit further in the distance, so that it actually looks like you aren't crashing into the sun?)

**Comment by Terin** [comment]
A star is just a planet right now, so that'd be a fairly significant refactor, yeah. It could also make the solar power/ramscoop effects not make as much sense, what with the dramatic rise you get when "on" a star that wouldn't really line up anymore.

**Comment by Ferociousfeind** [comment]
You don't get a very significant rise of either ramscoop or solar power when "on" a star, though! The game adds 1000 distance on top of your actual distance to the star _so that_ there is no significant boost to solar power or ramscoop, because you aren't parking on top of a star, you're still far above it.

**Comment by Terin** [comment]
"On top of" being relative. I figured you meant much further out, but if it's at something like 90-95%, that might work, which I mentioned at the top. Just need to figure out where that can go, and also if it'll look strange with ships landing on things when they're not at the center of the view.

**Comment by Ferociousfeind** [comment]
ohhh, that would be an issue, huh? There'd have to be some compensation in the "make ships get smaller until they land on the planet" code to make it look right away from the very center of the screen. I hadn't thought of that...

**Comment by gregorydgraham** [comment]
I like the effect overall and it seems to work pretty well.

I was a little disconcerted that the planets and star moved together, particularly when it was a ringworld and star.  Can the sun be set slightly further away than the objects?

**Comment by Terin** [comment]
Like I mentioned above, to the game right now, a star is basically a planet that you can't land on. It'd take some code to make it a distinct object that's drawn separately from planets. Even having planets drawn a bit separately is looking a little tough, as well.

**Comment by beccabunny** [comment]
> Like I mentioned above, to the game right now, a star is basically a planet that you can't land on. It'd take some code to make it a distinct object that's drawn separately from planets. Even having planets drawn a bit separately is looking a little tough, as well.

Perhaps it could be possible to make objects have a higher zoom the further they are from the center? So that the star have like 90% zoom, inner planets between 90 and 95, gas giants almost normal...
An issue I can see is that given that planet sprites are on a logarithmic scale that would be the opposite of what should realistically happen (excluding the stars),  but visually it could work

A probably better idea could be checking the sprite used by the object and applying the parallax if it is under the images/star/ subfolder, similarly to what happens with the landing messages (although that uses a whitelist, so this might actually be easier to work with in the data)

**Comment by m1m1k** [comment]
Amazing idea, and I love the look - I think it's definitely a step in the right direction.
This kind of improvement will move this game from "pretty decent looking" to "fantastic/AA".

**More ideas to add to it (for parallax v2)**
I'm also picturing some Asteroids/space junk also on another layers around the ship but above the planets (in some regions)
You could even have a few small ones go across the screen quickly above the ship (Possibly as black silhouettes) as is done in animation/cartoons for foreground parallax.
Slightly moving Dust/clouds/nebula in some systems would also add a lot too.

I'm harking back to an old "DnD" side-scroller arcade where layers of fog were in front of characters, also used in Chronotrigger.
https://lparchive.org/Chrono-Trigger-(by-Leavemywife)/Update%2033/15-image018.png
https://youtu.be/9Expd6rjzf0?t=848
but better/similar things have been done to add a haze effect (above the main sprites).
Just a little parallax and debri fields can add a LOT to making the game feel more 2.5D instead of 2D.
https://youtu.be/Ln-K25JUaHw?t=143   <-- an abandoned Star Trek game (no longer available), which basically was the same game as EV/Endless Sky

If some layers could be above/partly obscuring the ship, it could also be used as a fog of war/sensor disruption... ie - pirates and smugglers like to go through the nebula routes where they can't be seen or sensed.
There was also an ancient/crappy/awesome game called StarFlight II which had a nebula part of space where you couldn't see the star systems in it, so you had to wander around "in the dark" until you found them.  It was also a lot of fun, and great place for the bad guys/pirates/aliens to hide out and ambush unwary travelers.  Technically a spiritual ancestor to this game.
https://youtu.be/tg6FBlkrlQs?t=378   <-- Starflight II nebula
So, I picture using the parallax effect in Endless Sky to overlay some nebula/partly transparent clouds.

+1 vote for having different settings per system (I know it's hard/not possible right now), so that you get more variety in the game, and in different sectors of space. (the ember waste has a bit of dark/red theme to it, but since the UI never changes to reflect the different star cultures, more variance would also help make different parts of space "feel" differently (dark or light) and add a bit to the alien mystique of some places.

But until then, I'd say check in this great change! great idea!  way to go!

**Comment by Pointedstick** [comment]
Where are we with this?

**Comment by Zitchas** [comment]
From my understanding of comments on the discord, Terin is digging into this again, and looking into improve a few things:

From Terin on discord today:
> Been trying to crack back into refactoring parallax for layers. Going to need to figure out how to create and especially destroy layers on the fly...
> Yeah, I was able to hack together layers, but for being able to easily test them on the fly, and also supporting them in data (especially with the Ember Wastes vs. normal haze), it's some fairly significant refactor
At least, from where I am right now. Might not be too bad in the end, but I do need to get it working

**Comment by Zitchas** [comment]
Although that being said, @Terin Might it be beneficial to split up the project a bit? Right now this PR has the stars looking really good. Would it be worth getting this part merged just handling the stars, then work on other aspects as a follow-up? I seem to recall the haze looking pretty good most of the time too, although I seem to recall there being something off about it...

**Comment by Terin** [comment]
If people would be fine with it being merged simply as 1 layer of star, 1 layer of haze, and 1 layer of the rest, that'd be a-okay with me for starters. Breaking up stars/haze into even more layers enhances it even more, but requires some serious refactoring of how backgrounds are drawn, yeah.

**Comment by gregorydgraham** [comment]
Local haze, distant stars, and infinitely deep space seems like a reasonable number of layers

**Comment by Terin** [comment]
Alternatively, those extra layers could be hard-coded in, as well. The Ember Wastes couldn't have like, separate amounts of haze layers compared to normal, but it still looks pretty good as-is. The implementation would look even more hacky, which might be a deal-breaker for a tehhowch, but it could be worth looking into.

Might compile those videos again, ask around on the Discord about what should be merged from a visual perspective, and what should be merged from a code perspective.

**Comment by Terin** [comment]
I'm also slightly concerned that, with the new, amazing (but large) star sprites from #5828, the parallax effect might cause people to think the star should be more in the background. To make that possible from a code perspective, it would require even more serious refactoring, since stars/ships/planets are all drawn in the same way right now. And then if that happened, it would also beg gameplay questions of why a star in the background provides tons of solar energy when at like, a 90 degree angle, but a 100 degree angle causes a large drop.

I think it's still worth it to have it mostly as-is, but maybe a poll on the Discord could come in handy.

**Comment by Zitchas** [comment]
I can't speak to the code, but I wouldn't worry too much about the gameplay aspect - We all acknowledge the problems with the fact that we're operating in a massive 3d environment that has been flattened onto a 2d plane, and that entails discarding some of the aspects of reality that would normally be in play, such as the angles you mentioned there.

**Comment by Zitchas** [comment]
I just tried out Revamp again, and I like how it looks there. Having multiple layers of stars would definitely be nice and a worthwhile thing to work towards, but simply having it the way it is there is a big graphical improvement as-is. So two, having a "main" haze layer that is underneath the ship with a secondary (and thinner) haze overtop of the ship layer would be awesome, but again I think that would be a great future work rather than something that must be in the current PR.

I don't know how it would look from your point of view, knowing the code. From my conceptual point, I'd imagine something like.
1. the current PR with just a few layers.
2. Spreading out the stars into a number of disparate layers. (I seem to recall testing a version of parallax that did this. Didn't I?)
3. Enabling multiple haze layers, including one that is above the ship layer

**Comment by Terin** [comment]
Looking for a temperature check on this: Would this implementation (three layers of zoom, 100% for sprites, 90% for haze, 70% for stars) be possible for merging, and then future PRs could add support for sub-layers and data integration? Or would it need those, combined with all the refactored classes and the like, to be considered? Like Zitchas said above, while further layers would look nice, I think the effect on its own is a pretty marked improvement, and would like to see it in game as-is, if possible.

**Comment by Zitchas** [comment]
I would be happy with the three layers being merged as a great starting point.

**Comment by Amazinite** [comment]
This sounds like a good base for now. I'll give it a look later.
Since v0.9.14 still isn't out on Steam, I'll be stacking a few more things onto the v0.9.15 milestone. Little use in releasing a major update when the most recent minor one isn't even widely available yet.

File: source/StarField.cpp [hunk 1]
```diff
@@ -119,9 +122,14 @@ void StarField::SetHaze(const Sprite *sprite, bool allowAnimation)
 
 void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 {
+	double savedZoom = zoom;
```
Comment (Amazinite) [inline source/StarField.cpp:125]: I think a term like "baseZoom" would be better.

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/StarField.cpp [hunk 2]
```diff
@@ -42,6 +42,9 @@ namespace {
 	const size_t HAZE_COUNT = 16;
 	// This is how fast the crossfading of previous haze and current haze is
 	const double FADE_PER_FRAME = 0.01;
+	// Percentage of screen the stars/haze are from the screen, to simulate parallax
```
Comment (Amazinite) [inline source/StarField.cpp:45]:
```suggestion
	// An additional zoom factor applied to stars/haze on top of the base zoom, to simulate parallax.
```

File: source/StarField.cpp [hunk 3]
```diff
@@ -119,9 +122,14 @@ void StarField::SetHaze(const Sprite *sprite, bool allowAnimation)
 
 void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 {
+	double baseZoom = zoom;
+
 	// Draw the starfield unless it is disabled in the preferences.
 	if(Preferences::Has("Draw starfield"))
 	{
+		// Modify zoom for first parallax layer
```
Comment (Amazinite) [inline source/StarField.cpp:130]:
```suggestion
		// Modify zoom for the first parallax layer.
```

File: source/StarField.cpp [hunk 4]
```diff
@@ -179,6 +187,10 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 	if(!Preferences::Has("Draw background haze"))
 		return;
 	
+	// Modify zoom for second parallax layer
```
Comment (Amazinite) [inline source/StarField.cpp:190]:
```suggestion
	// Modify zoom for the second parallax layer.
```

**Review by Amazinite — COMMENTED** [review]
Smh make your comments full sentences.

**Pull request merged**